### ISSUE TITLE
Weak reference

### ISSUE THREAD

**Issue by robfors** [mruby/mruby#4162]
I would like to save weak references to ruby objects of type MRB_TT_DATA.
I have been trying to save the object's `mrb_value` but I have found that, upon reusing the `mrb_value` later, I get unexpected behaviour such as raising NoMethodError when calling a method on the object that I know exists.
So my question is, should I be able to safety store an `mrb_value` and use it well later in the program as long as the gc callback has not been called on the object?
If not, can I use something like `mrb_ptr()` to store some sort of pointer to it?

**Comment by dmajkic** [comment]
It looks like your host language environment (hosting mruby) destroyed object whose reference was in MRB_TT_DATA. Probably that language GC decided that there are no references to object, and collected it. 

If that was the case - try to signal to host language GC that object is still "alive" while in MRB_TT_DATA.

**Comment by matz** [comment]
Weak references cannot be implemented without associating garbage collector update. Currently, we have no plan to tweak GC to implement weak references.
Reopen the issue if you really really need weak references (at our cost).

**Comment by robfors** [comment]
What I am building is a project to access java script objects in ruby (and vice versa but I have that working successfully). Every time I call a js function that returns a js object it will be converted to a wrapper (lets call them a `JSWrapper`). A `JSWrapper` stores a reference to it's js object as a pointer to an `emscripten::val`. For this reason `JSWrapper` are of type MRB_TT_DATA. This part works well, but if I call the function again and it returns the same js object I want it to be converted into the same `JSWrapper` instance that is already pointing to that js object. For this reason, when a new `JSWrapper` is built I take it's `mrb_value`, store it as a property of the js object and can supposedly look for and retrieve that property every time a js object is converted in to a `JSWrapper` that has already been previously converted.
My assumption is, because `JSWrapper` is already of type MRB_TT_DATA, when the gc callback is called I can destroy the js property and therefor `mrb_value` from being used accidentally in the future. Is this assumption correct? My bug my be related to how I am store and retrieving the `mrb_value` but ether way I wanted to ensure that this was possible. Weak references are important, but for this use I really don't think any modifications need to be made to _mruby_.

**Comment by dmajkic** [comment]
@robfors Try this: when you store object in JSWrapper, store refference to it in a JS hash, or JS array on emscripten side. Check if then your object gets deleted. If it works that way, your object is probably GC-ed on empscripten side, as it "thinks" that there are no live references on JS side to that object.

**Comment by robfors** [comment]
@dmajkic thanks for the suggestion, I gave that a try, unfortunately I am still having the the issue.

Something weird that I am seeing is duplicate addresses in use.
I have found that I have a faulty assumption on how _mruby_ works. To experiment, I was calling something like 
```cpp
  mrb_value v;
  v = mrb_obj_new(mrb, class_mrb, 0, NULL);
  printf("v%p\n", (void *)&mrb_ptr(v));
  v = mrb_obj_new(mrb, class_mrb, 0, NULL);
  printf("v%p\n", (void *)&mrb_ptr(v));
```
expecting that the pointer of each object would be different or that the gc callback would be called, but instead it seems to simply overwrite the rb object, so now I am confused.

Here is a simplified factory function that builds `JSWrapper`s.
```cpp
mrb_value build(val object_js)
{
  mrb_value js_wrapper;
  
  // try to get existing js_wrapper
  if (object_js_has_backward_reference_property(object_js))
  {
    void* js_wrapper_ptr = get_js_wrapper_ptr(object_js);
    js_wrapper = mrb_obj_value(js_wrapper_ptr);
    return js_wrapper;
  }
  js_wrapper = mrb_obj_new(mrb, class_mrb, 0, NULL);

  // set up forward reference (this part works)
  //   (rb:JSWapper -> c:mrb_value -> cpp:val -> js:object)
  val* object_js_ptr = (val*)mrb_malloc(mrb, sizeof(val));
  if (object_js_ptr == nullptr)
    // throw OutOfMemoryError
  new (object_js_ptr) val(object_js);
  mrb_data_init(js_wrapper, object_js_ptr, &_internal_data_type);

  // set up backward reference
  //    (js:object -> js:rb-weak-reference -> c:mrb_value -> rb:JSWapper)
  void* js_wrapper_ptr = mrb_ptr(js_wrapper);
  set_js_wrapper_ptr(js_wrapper_ptr, object_js);
  
  return js_wrapper;
}
```
Am I overwriting existing rb objects like I was in my experiment? Im not really sure how to fix it.

**Comment by robfors** [comment]
I have narrowed down the issue. I believe the _mruby_ gc is collecting an object that is still referenced.
I am calling this rb method (from ruby):
```ruby
def array(array_rb)
  # array_rb == [#<JSWapper>, #<JSWapper>, #<JSWapper>] # the first element is the one that gets mistakenly collected soon
  array_js = JSWapper.global.call(JSWapper.string("Array"))
  array_rb.each { |item| array_js.call(JSWapper.string("push"), item) }
  array_js
end
```

- `JSWapper::global` is a c function that returns the element that will get mistakenly collected soon
- `array_js` is a `JSWapper`
- `JSWapper#call` is a c function
-  `JSWapper::string` is also a c function
-  the gc is arbitrary run during `JSWapper.string("push")` of the first iteration of each
-  by the time the first array_js.call is run, `item` has already been mistakenly destroyed 

Could there be an _mruby_ bug where _mruby_ sees the result returned from `JSWapper::global` and thinks that, because it is not referenced any more (and some how ignore the reference from `array_rb`) it can be destroyed?

I don't know if there is much here to go on. I wish I could simplify my code base so I could submit something so the problem could be reproduced, but the bug is partially non deterministic, every change I make (even after the point of execution of the error) will change it's behaviour. Sometimes my test suite fully passes and sometimes I get other random errors. I can say, at least, that the bug goes away when I disable the gc.

**Comment by matz** [comment]
Is it possible to investigate how `JSWrapper` is implemented (by source code)?

**Comment by robfors** [comment]
That would be awesome. I have created some git repos.
Note that `JSWapper` is actually called `JSInterface` in my code.
Look at [js_interface.cpp](https://github.com/robfors/bruby-bridge-mruby-embind/blob/dev/src/js_interface.cpp) and [js_interface.hpp](https://github.com/robfors/bruby-bridge-mruby-embind/blob/dev/src/js_interface.hpp) for comments.

Here is the way my project is structured:
[bruby-bridge-mruby-embind](https://github.com/robfors/bruby-bridge-mruby-embind/tree/dev) implements the low level stuff, specific to _mruby_ and _embind_. Look at the _dev_ branch.
[bruby-bridge](https://github.com/robfors/bruby-bridge) is built on _bruby-bridge-mruby-embind_. It is `.rb` and `.js` files only. It has the `JSValue` class which is built on `JSInterface`.

Running the test suite of _bruby-bridge-mruby-embind_ unfortunately passes with no issues. It's only when I run the test suite of _bruby-bridge_ the the bug pops up, but given that  _bruby-bridge_ is only `.rb` files it obviously does not have any non deterministic bugs.

The method where the bug happens (at the moment) is at [interface_array](https://github.com/robfors/bruby-bridge/blob/debug/mrblib/bruby_bridge/js_method.rb#L207).

Because this project compiles to js code, _gdb_ is out of the question. In the _debug_ branch of _bruby-bridge-mruby-embind_ I have added a bunch of `printf`s to track the execution. In [log.txt](https://github.com/robfors/bruby-bridge/blob/debug/test/log.txt) I have logged the output of running the test suite of _bruby-bridge_ (note the bottom of it, I have added some of the js objects the JSInterfaces are referring to and it shows the point in execution that the gc collected the object). Note that it only made it to 13 of the 100 long loop at the top of [js_value.rb](https://github.com/robfors/bruby-bridge/blob/debug/test/lib/js_value.rb).

If you need to run it yourself let me know and I will set everything up and give instructions on how to get to this result.

Thanks for any help you can give to solve this mess!

**Comment by robfors** [comment]
If you think my implementation of `JSInterface` is correct and their might be a bug with _mruby_s gc, I could replace the `emscripten::val` with a dummy class and compile everything to a traditional assembly executable. Then _gdb_ could be used to search for the bug. Would that help?

**Comment by robfors** [comment]
I finally found the culprit! It was the overriding of `==` and  `!=` (either one will break things). I guess _mruby_ calls these internally and can not handle these methods doing complex operations in C (or being overridden at all?). I tried modifying them to immediately return the obvious value if the other operand is not of type `JSInterface`, and then simply returning `true` or `false`, but it still results in memory corruption.

As far as my project is concerned, leaving the default comparators should work for some of the cases, just not comparing JavaScript primitives as they are always converted into new `JSInterface`s regardless if one already exists.

I guess it was _Ruby_ code that produced non deterministic behaviour!
